docs/ddb.1
.TH DDB 1 "17 February 2018" "GLADSERV" "LOCAL USER COMMANDS"
.SH NAME
ddb \- distributed device backup
.SH SYNOPSIS
.EX
\fBddb\fP [options]... \fISOURCE\fP [\fIDESTINATION\fP]
.EE
.SH DESCRIPTION
The purpose of the
.B ddb
utility is to copy data between block devices, or between entities
which look like block devices; it is similar
.BR dd (1)
command with some important differences which make it more suitable for
use in distributed backup situations; it also works well in emergency
rescue situations in which the source device may be damaged.
.LP
One difference between \fBdd\fP abd \fBddb\fP is that
if a block cannot be read,
.B dd
will offer the choice to stop the copy or replace the block with zeros,
while
.B ddb
will keep trying until the hardware manages to correct the error or too
may retries are reached; and even in this case, re-running
.B ddb
will continue trying to recover the missing blocks. In a distributed
setup, a copy error can be the result of network issues, so it is
always necessary to retry operations which failed.
.LP
Another important difference between
.B dd
and
.B ddb
is that
.B ddb
will not write to
.I DESTINATION
unless it needs to. This means that if
.I DESTINATION
is a file it will be a "sparse" file, in which blocks containing only
zeroes are not allocated. The result will work exactly the same as a
normal file, but use less space. Also, if
.I DESTINATION
is a solid state disk or other medium allowing a limited number of writes,
.B ddb
will cause as little wear and tear as possible, and running
.B ddb
several times will result in no additional wear and tear.
.LP
Finally, \fBddb\fP extends the concept of a "device" to include direct
access to remote and emulated devices, and allows the use of plugins
to extend the concept of "device" further. One important example of
a "device" which isn't a traditional block device or an image file is
the "sequence" which contains a full copy of a device and any number
of incremental copies; one could use this type of device as the
destination for a backup, and if a restore is required this can be
done using any of the previous versions present in the sequence. See
.BR ddb-sequence (7)
and
.BR ddb-sequence (1)
.LP
The \fIDESTINATION\fP can be omitted from the command line; in
this case, the program reads the \fISOURCE\fP and checks that it
is completely readable, but of course no actual copy is performed.
.LP
.SH OPTIONS
Normally, just running the program without any special options:
.LP
.EX
\fBddb\fP \fISOURCE\fP \fIDESTINATION\fP
.EE
.LP
will select reasonable defaults and copy the data. A file in a configuration
directory will specify any special handling required by \fISOURCE\fP
and/or \fIDESTINATION\fP, as described in
.BR ddb (5)
however a number of options control how the program runs if it is
desired to change the defaults.
.LP
.TP 0.5i
.B \-b \fIBLOCK_SIZE\fP
The copy reads a number of blocks, of size \fIBLOCK_SIZE\fP bytes each,
from \fISOURCE\fP, then checks which of these need to be written to
\fIDESTINATION\fP and writes them. This option changes the default
\fIBLOCK_SIZE\fP. For best results, this needs to be a multiple of
the internal block size of both \fISOURCE\fP and \fIDESTINATION\fP,
and some types of device will only work with some block sizes. The
default is 4096.
.TP 0.5i
.B \-c [\fIINTERVAL\fP:]\fIFILE\fP
The program will write its internal state to \fIFILE\fP every
\fIINTERVAL\fP seconds (default 60). If this file exists when
the program starts, and it contains a valid state for the copy,
the program will automatically resume from that state rather
than running the full copy (if the file exists, but it is not
a valid state the program will refuse to run, to avoid overwriting
something which isn't a state file). If this file does not exist
when the program starts, it will create one.
.TP 0.5i
.B \-C
Do not use checksums to determine whether two blocks are identical;
checksums can save time when one or both of \fISOURCE\fP or
\fIDESTINATION\fP are remote because they use a lot less network
bandwidth than the full data; they are automatically disabled if
both devices are local.
.TP 0.5i
.B \-f \fIFILE\fP
At end of processing, writes to \fIFILE\fP the list of blocks which
have been successfully copied to the \fIDESTINATION\fP or which the program
determined did not need to be copied.  The file format is the same as the
one used by
.B \-i
and
.B \-o.
If there is no
.I DESTINATION
this file records which blocks can be read from the \fISOURCE\fP.
.TP 0.5i
.B \-F \fIINTERVAL\fP
Flushes pending writes to \fIDESTINATION\fP every \fIINTERVAL\fP
seconds; if \fIINTERVAL\fP is zero, disables this; default is 120
seconds
.TP 0.5i
.B \-h
Print an usage message with some of the information in this manual page;
then exit the program.
.TP 0.5i
.B \-i \fIFILE\fP
Reads the list of blocks to copy from \fIFILE\fP which normally
has been produced using
.B \-o.
.TP 0.5i
.B \-k \fIDIRECTORY\fP
Replaces the default user configuration directory with \fIDIRECTORY\fP;
if this option is repeated, the program will search all the directories
specified (but not the default, which can be added again if desired
by specifying this option one more time). If the \fIDIRECTORY\fP is not
absolute it will be relative to $HOME. To completely remove the user
configuration directories specify a single \fB\-k\-\fP.
.TP 0.5i
.B \-K \fIDIRECTORY\fP
Replaces the default system configuration directory with \fIDIRECTORY\fP;
if this option is repeated, the program will search all the directories
specified (but not the default, which can be added again if desired
by specifying this option one more time). To completely remove the
system configuration directories specify a single \fB\-K\-\fP.
.TP 0.5i
.B \-l
Print the program's licence on standard output and exit.
.TP 0.5i
.B \-n
Do not make any changes to
.I DESTINATION
but just report if it differs from \fISOURCE\fP.  This switch implies
.BR \-p\fI1\fP
(see below).
.TP 0.5i
.B \-o \fIFILE\fP
Writes the list of blocks which could not be recovered to \fIFILE\fP.
This can be used in various ways, as described in the examples.
.TP 0.5i
.B \-O
If \fB\-o\fP is specified, write this file at the end of every pass instead
of just at the end of the last pass.
.TP 0.5i
.B \-q
Quiet mode: only produce error messages and some summaries. Without this
option the program produces a progress report every few seconds.
.TP 0.5i
.B \-p \fIMAX_PASS\fP
Determine the maximum number of "passes" (retries after the copy
encouters some read errors). The default is 10. The program will
first copy
.I SOURCE
to
.I DESTINATION
skipping any blocks containing uncorrectable errors, but keeping a list
of such blocks in memory (and the state file, if \fB\-c\fP is specified).
This is the first pass. If the number of passes is greater than 1,
.B ddb
will then retry just the blocks which had errors, again keeping a list
of the ones which failed a second time. The process is repeated until the
whole of the
.I SOURCE
has been correctly copied, or the maximum number of passes is reached.
If there are still uncorrected read errors at the end, the program will
print a warning to this effect, and if the
.BR \-o
option was specified, puts the list of uncorrected blocks in that file.
Run the program again to keep retrying.
.TP 0.5i
.B \-P [\fIINTERVAL\fP:]\fIFILENAME\fP
Writes a program-readable progress report to \fIFILENAME\fP every
\fIINTERVAL\fP seconds (default 120); see
.BR ddb_copy (3)
for the format
.TP 0.5i
.B \-r \fIINTERVAL\fP
Changes the time between reports to the interval specified (default: 2
seconds). If \fIINTERVAL\fP is zero, no progress report is provided.
.TP 0.5i
.B \-R
Provides an extra progress report at the end of each pass, showing the
final number of blocks read, copied, skipped, etc. This report can be
useful to check that the program is copying the amount of data one expects.
.TP 0.5i
.B \-s \fIINTERVAL\fP
Sleep the specified time interval after each progress report (this sleep
happens even if the progress report is silenced by using the
.B \-q
option).
.TP 0.5i
.B -S \fINUMBER\fP
Specify the number of blocks copied in a single operation; default is
32, and the value can be between 1 and 1024; larger values decrease the
effect of network latency on remote devices, but increase memory usage;
for local devices, lower values work best as they reduce unnecessary buffers.
.TP 0.5i
.B -t \fITYPE\fP
Specify a device \fTYPE\fP for \fIDESTINATION\fP. The default is to
autodetect the type of existing devices, and to infer a type from the
device name if writing to something which does not yet exist. When this
option is provided for existing devices, the program will refuse to run
if the device is not of the correct type. When running
.BR ddb
from a script it is always best to specify this option.
.TP 0.5i
.B -T \fITYPE\fP
Specify a device \fTYPE\fP for \fISOURCE\fP. The default is to autodetect
the type. When this option is provided, the program will refuse to run if
the device is not of the correct type. When running
.BR ddb
from a script it is always best to specify this option.
.TP 0.5i
.B \-v
Print version number and exit.
.TP 0.5i
.B \-x
Ask that \fIDESTINATION\fP is not present when the program starts and
therefore will be created by the program ("e\fBx\fPclusive" mode).
.TP 0.5i
.B \-w
Write all blocks to \fIDESTINATION\fP even if they are already present
and identical to the data which will be written.
.PP
.LP
.SH EXAMPLES
To create a full backup from a device into a new "sequence":
.EX
\fBddb\fP \fB\-x\fP \fP\-t\fP sequence \fIDEVICE\fP /var/backups/\fIDEVICE\fP
.EE
.LP
To add an incremental backup to an existing "sequence":
.EX
\fBddb\fP \fP\-t\fP sequence \fIDEVICE\fP /var/backups/\fIDEVICE\fP
.EE
.LP
To restore the device to the state it had the last time it was backed up:
.EX
\fBddb\fP \fP\-T\fP sequence /var/backups/\fIDEVICE\fP \fIDEVICE\fP
.EE
.LP
And to restore to the state it had at the time of the most recent backup
on or before December 1, 2017:
.EX
\fBddb\fP \fP\-T\fP sequence /var/backups/\fIDEVICE\fP/2017-12-01 \fIDEVICE\fP
.EE
.LP
To try to rescue data from \fIDEVICE\fP trying 1000 times if that's what
it takes, and saving the program's state every 2 minutes so that the
copy can be resumed if it gets interrupted:
.LP
.EX
\fBddb\fP \fB\-p\fP 1000 \fB\-c\fP /var/tmp/ddb-state \fIDEVICE\fP /var/tmp/rescued
.EE
.LP
The above can also be rewritten without saving the state as:
.LP
.EX
\fBddb\fP \fB\-o\fP /var/tmp/ddb-retry \fB\-O\fP \fIDEVICE\fP /var/tmp/rescued
.EE
.LP
If the program is interrupted, the file /var/tmp/ddb-retry will contain
the list of blocks which still needed to be copied at the end of the
last pass it completed, so the copy can be resumed with:
.EX
\fBddb\fP \fB\-o\fP /var/tmp/ddb-retry \fB\-O\fP \fB\-i\fP /var/tmp/ddb-retry \fIDEVICE\fP /var/tmp/rescued
.EE
.LP
Get a list of blocks in \fIDEVICE\fP which have read errors:
.EX
\fBddb\fP \fB\-p\fP 1 \fB\-o\fP /var/tmp/bad-blocks \fIDEVICE\fP
.EE
.LP
Now try to copy these blocks:
.EX
\fBddb\fP \fB\-P\fP 1000 \fB\-i\fP /var/tmp/bad-blocks \fIDEVICE\fP /var/tmp/rescued
.EE
.LP
If all these blocks can eventually be read, one can try writing them
back to the device which will reallocate the blocks to a new area to
avoid the errors; we use \fB\-w\fP to write unconditionally, and also
add a delay to give the device a better chance to catch up:
.EX
\fBddb\fP \fB\-i\fP /var/tmp/bad-blocks \fB\-w\fP \fB\-s\fP 2 /var/tmp/rescued \fIDEVICE\fP
.EE
.SH "EXIT STATUS"
.B ddb
exits with a status of zero if all data was successuflly copied from
.I SOURCE
to \fIDESTINATION\fP.
.LP
An exit status of 1 signals an error in the options provided or a
problem initialising the program (there will be a message explaining
what went wrong).
.LP
An exit status of 2 signals an error during processing, for example
an error updating
.I DESTINATION.
The program will produce a suitable error message before exiting.
.LP
An exit status of 3 signals that some uncorrected read errors still
exist in
.I SOURCE
but the maximum number of passes was reached. A message will say how
many errors remain and lists the corresponding block numbers.
.SH ENVIRONMENT
When starting,
.B ddb
reads the following environment variables, if present:
.TP 0.5i
.I DDB_CONFIG
Location of default user configuration file, instead of $HOME/.ddb;
the \fB\-k\fP option will override this.
.TP 0.5i
.I DDB_SYSCONFIG
Location of default user configuration file, instead of /etc/ddb;
the \fB\-K\fP option will override this.
.TP 0.5i
.I DDB_MODPATH
Location of loadable modules (plugins), overriding compiled-in
defaults.
.PP
.SH FILES
.nf
$HOME/.ddb                   Default user configuration directory
/etc/ddb                     Default system configuration directory
.fi
.SH "SEE ALSO"
.BR dd (1),
.BR ddb-info (1),
.BR ddb-sequence (1),
.BR ddb_copy (3),
.BR ddb (5),
.BR ddb-sequence (7)
.SH BUGS
None that we know of. If we do become aware of a bug, we'll work
to fix it.
.SH COPYRIGHT
Copyright (c) 2018 Claudio Calvelli <ddb@gladserv.com>
.LP
All rights reserved.
.LP
Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:
.TP 0.5i
1.
Redistributions of source code must retain the above copyright
notice, this list of conditions and the following disclaimer.
.TP 0.5i
2.
Redistributions in binary form must reproduce the above copyright
notice, this list of conditions and the following disclaimer in the
documentation and/or other materials provided with the distribution.
.TP 0.5i
3.
If the program is modified in any way, a line must be added to the
above copyright notice to state that such modification has occurred.
.LP
THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS "AS IS" AND
ANY EXPRESSED OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS
BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
THE POSSIBILITY OF SUCH DAMAGE.


src/ddb.c
/* ddb.c - distributed device backup
 *
 * Copyright (c) 2018 Claudio Calvelli <ddb@gladserv.com>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. If the program is modified in any way, a line must be added to the
 *    above copyright notice to state that such modification has occurred.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS \"AS IS\" AND
 * ANY EXPRESSED OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
 * THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <limits.h>
#include <string.h>
#include <memory.h>
#include <ctype.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <time.h>
#include <endian.h>
#include <ddb.h>
#include <ddb-private.h>

#define DEFAULT_CHECKPOINT_INTERVAL 60
#define DEFAULT_MACHINE_PROGRESS_INTERVAL 120
#define DEFAULT_PROGRESS_INTERVAL 2
#define DEFAULT_FLUSH_INTERVAL 120
#define DEFAULT_MAX_PASSES 10

static ddb_copy_t W;
static const char * src_type = NULL;
static const char * dst_type = NULL;
static int default_copy_block = 0;
static int clear_user_cfg = DDB_CONFIG_CLEAR;
static int clear_system_cfg = DDB_CONFIG_CLEAR;
static int dst_exclusive = 0;
static int device_report = 0;
static int do_help = 0;
static int do_licence = 0;
static int do_version = 0;

static void init_copy(void) {
    W.src_name = NULL;
    W.src = NULL;
    W.dst_name = NULL;
    W.dst = NULL;
    W.write_dst = 1;
    W.total_size = 0;
    W.total_blocks = 0;
    W.block_size = DEFAULT_BLOCK_SIZE;
    W.input_list = NULL;
    W.max_passes = DEFAULT_MAX_PASSES;
    W.checkpoint_file = NULL;
    W.checkpoint_interval = DEFAULT_CHECKPOINT_INTERVAL;
    W.progress_function = ddb_progress_print;
    W.progress_arg = stdout;
    W.progress_interval = DEFAULT_PROGRESS_INTERVAL;
    W.machine_progress_file = NULL;
    W.machine_progress_interval = 0;
    W.extra_report = 0;
    W.output_list = NULL;
    W.output_each_pass = 0;
    W.flush_interval = DEFAULT_FLUSH_INTERVAL;
    W.use_checksums = 1;
    W.skip_identical = 1;
    W.copied_list = NULL;
}

/* checkpoint header on disk */
typedef struct {
    int64_t magic;
    int64_t total_size;
    int64_t blocks_read;
    int64_t read_errors;
    int64_t blocks_written;
    int64_t blocks_skipped;
    int64_t write_errors;
    int32_t block_size;
    int32_t pass;
} __attribute__((packed)) checkpoint_header_t;

#define CHECKPOINT_MAGIC 0x43686b506f696e74LL /* "ChkPoint" */

static void usage(FILE * F) {
    fprintf(F,
"Usage: %s [OPTIONS] SOURCE [DESTINATION]\n"
"-b BLOCK_SIZE   Change block size from default %d\n"
"-c [INTERVAL:]CHECKPOINT\n"
"                Checkpoint file: see documentation\n"
"-C              Do not use checksums to determine if some data is already\n"
"                present on DESTINATION, instead compare the full data\n"
"-e              Provides an extra per-device report at end, if supported\n"
"-f FILENAME     At end of processing, write to FILENAME the list of blocks\n"
"                which were copied from SOURCE to DESTINATION or which were\n"
"                already present in DESTINATION; if there is no DESTINATION\n"
"                the list of blocks which could be read from SOURCE\n"
"-F INTERVAL     Flush data to DESTINATION every INTERVAL seconds; use\n"
"                \"-F 0\" to disable; default is %d seconds\n"
"-h              Print this helpful message and exit\n"
"-i FILENAME     Before starting, read list of blocks to copy from FILENAME;\n"
"                default is to copy all blocks present in SOURCE\n"
"-k DIRECTORY    Overrides default user configuration directory\n"
"                ($HOME/%s or if defined $%s)\n"
"-K DIRECTORY    Overrides default system configuration directory\n"
"                (%s or if defined $%s)\n"
"-l              Print program's licence and exit\n"
"-n              Open DESTINATION readonly and report what would be written to\n"
"                it; this is incompatible with \"-w\"; implies \"-p 1\"\n"
"-o FILENAME     At end of processing, write the list of blocks which could not\n"
"                be copied to FILENAME\n"
"-O              With \"-o\", write the list after each pass\n"
"-p PASSES       Maximum number of passes (retries) before a block is considered\n"
"                completely unreadable; default is %d\n"
"-q              Omit per-pass and progress messages (but still obey \"-s\")\n"
"-r INTERVAL     Interval between progress reports, default is %d seconds;\n"
"                use \"-r 0\" to disable progress reports\n"
"-R              Show an extra progress report before the \"end pass\" message\n"
"-s INTERVAL     Sleep INTERVAL seconds after progress reports; default 0\n"
"-S NUMBER       Size of the copy buffer, default %d\n"
"-t TYPE         If creating DESTINATION, make it with type TYPE (see\n"
"                documentation); if it already exist, check it has this type\n"
"-T TYPE         Specify that SOURCE has the given TYPE, if autodetection fails\n"
"-v              Print program's version information and exit\n"
"-x              Exclusive: if DESTINATION already exist, stop with an error\n"
"-w              Write unconditionally to DESTINATION even if the data is\n"
"                already resent and identical; incompatible with \"-n\";\n"
"                see documentation about using this option\n"
	    , progname, DEFAULT_BLOCK_SIZE, DEFAULT_FLUSH_INTERVAL,
	    ddb_default_config(), ddb_override_config(),
	    ddb_default_sysconfig(), ddb_override_sysconfig(),
	    DEFAULT_MAX_PASSES, DEFAULT_PROGRESS_INTERVAL, default_copy_block);
}

static int open_devices(void) {
    ddb_device_info_t info;
    W.src = ddb_device_open(W.src_name, src_type,
			    (size_t)W.block_size, O_RDONLY, (off_t)0);
    if (! W.src) {
	perror(W.src_name);
	return 0;
    }
    ddb_device_info(W.src, &info);
    W.total_size = info.total_size;
    W.total_blocks = info.num_blocks;
    if (W.dst_name) {
	int flags;
	if (W.write_dst) {
	    flags = W.skip_identical ? O_RDWR : O_WRONLY;
	    flags |= O_CREAT;
	    if (dst_exclusive) flags |= O_EXCL;
	} else {
	    flags = O_RDONLY;
	}
	W.dst = ddb_device_open(W.dst_name, dst_type,
				(size_t)W.block_size, flags, W.total_size);
	if (! W.dst) {
	    perror(W.dst_name);
	    return 0;
	}
    }
    return 1;
}

static int send_line(const char * line, void * arg) {
    puts(line);
    return 0;
}

static int close_devices(void) {
    if (device_report) {
	ddb_device_report(W.src, send_line, NULL);
	if (W.dst) ddb_device_report(W.dst, send_line, NULL);
    }
    ddb_device_close(W.src);
    if (W.dst && ddb_device_close(W.dst) < 0) {
	perror(W.dst_name);
	return 0;
    }
    return 1;
}

#define store_int(var, min, max) \
    if (! ddb_store_int(opt, optarg, &var, min, max)) return 0
#define store_cfg(clr, which) \
    if (! ddb_store_cfg(&clr, which)) return 0
static int parse_arguments(int argc, char * argv[]) {
    int opt, narg, rwmax;
    init_copy();
    opterr = 0;
    while ((opt = getopt(argc, argv,
			 ":b:c:Cdf:F:hi:k:K:lno:Op:P:qr:Rs:S:t:T:vxw")) != -1)
    {
	char * colon;
	switch (opt) {
	    case '?' :
		fprintf(stderr, "%s: invalid option \"-%c\"\n",
			progname, argv[optind - 1][1]);
		usage(stderr);
		return 0;
	    case ':' :
		fprintf(stderr, "%s: option \"-%c\" requires an argument\n",
			progname, argv[optind - 1][1]);
		return 0;
	    case 'b' :
		store_int(W.block_size, MIN_BLOCK_SIZE, MAX_BLOCK_SIZE);
		break;
	    case 'c' :
		colon = index(optarg, ':');
		if (colon) {
		    int l = colon - optarg, i, ok = l > 0;
		    char tmp[l + 1];
		    for (i = 0; i < l; i++) {
			if (! isdigit(optarg[i])) {
			    ok = 0;
			    break;
			}
			tmp[i] = optarg[i];
		    }
		    if (ok) {
			tmp[l] = 0;
			if (! ddb_store_int(opt, tmp, &W.checkpoint_interval,
					    5, INT_MAX))
			    return 0;
			optarg = colon + 1;
		    }
		}
		W.checkpoint_file = optarg;
		break;
	    case 'C' :
		W.use_checksums = 0;
		break;
	    case 'd' :
		device_report = 1;
		break;
	    case 'f' :
		W.copied_list = optarg;
		break;
	    case 'F' :
		store_int(W.flush_interval, 0, INT_MAX);
		break;
	    case 'h' :
		do_help = 1;
		break;
	    case 'i' :
		W.input_list = optarg;
		break;
	    case 'k' :
		store_cfg(clear_user_cfg, DDB_CONFIG_USER);
		break;
	    case 'K' :
		store_cfg(clear_system_cfg, DDB_CONFIG_SYSTEM);
		break;
	    case 'l' :
		do_licence = 1;
		break;
	    case 'n' :
		W.write_dst = 0;
		break;
	    case 'o' :
		W.output_list = optarg;
		break;
	    case 'O' :
		W.output_each_pass = 1;
		break;
	    case 'p' :
		store_int(W.max_passes, 1, INT_MAX);
		break;
	    case 'P' :
		colon = index(optarg, ':');
		if (colon) {
		    int l = colon - optarg, i, ok = l > 0;
		    char tmp[l + 1];
		    for (i = 0; i < l; i++) {
			if (! isdigit(optarg[i])) {
			    ok = 0;
			    break;
			}
			tmp[i] = optarg[i];
		    }
		    if (ok) {
			tmp[l] = 0;
			if (! ddb_store_int(opt, tmp,
					    &W.machine_progress_interval,
					    5, INT_MAX))
			    return 0;
			optarg = colon + 1;
		    }
		} else {
		    W.machine_progress_interval =
			DEFAULT_MACHINE_PROGRESS_INTERVAL;
		}
		W.machine_progress_file = optarg;
		break;
	    case 'q' :
		W.progress_function = NULL;
		break;
	    case 'r' :
		store_int(W.progress_interval, 0, INT_MAX);
		break;
	    case 'R' :
		W.extra_report = 1;
		break;
	    case 's' :
		store_int(W.progress_sleep, 0, INT_MAX);
		break;
	    case 'S' :
		store_int(rwmax, 1, INT_MAX);
		ddb_copy_block(rwmax);
		break;
	    case 't' :
		dst_type = optarg;
		break;
	    case 'T' :
		src_type = optarg;
		break;
	    case 'v' :
		do_version = 1;
		break;
	    case 'x' :
		dst_exclusive = 1;
		break;
	    case 'w' :
		W.skip_identical = 0;
		break;
	}
    }
    if (do_licence || do_version || do_help) return 1;
    narg = argc - optind;
    if (narg < 1 || narg > 2) {
	usage(stderr);
	return 0;
    }
    W.src_name = argv[optind];
    if (narg > 1) W.dst_name = argv[optind + 1];
    if (W.skip_identical == 0 && W.write_dst == 0) {
	fprintf(stderr, "%s: cannot use both \"-w\" and \"-n\"\n", progname);
	return 0;
    }
    if (! W.skip_identical) W.use_checksums = 0;
    if (W.dst && ! W.write_dst) W.max_passes = 1;
    return 1;
}

int main(int argc, char * argv[]) {
    int ok;
    progname = rindex(argv[0], '/');;
    if (progname) progname++; else progname = argv[0];
    default_copy_block = ddb_copy_block(0);
    if (! parse_arguments(argc, argv)) return 1;
    if (do_version || do_licence) {
	printf("ddb %s\n", ddb_version);
	if (do_licence) putchar('\n');
    }
    if (do_help) usage(stdout);
    if (do_licence)
	puts(ddb_licence);
    if (do_licence || do_version || do_help)
	return 0;
    if (! open_devices()) return 2;
    ok = ddb_copy(&W);
    if (ok < 0) return 2;
    if (! close_devices()) return 2;
    return ok ? 0 : 3;
}

